### ISSUE TITLE
Spoofing by patching the target app requiring GCM, instead of microG itself?

### ISSUE THREAD

**Issue by ccaapton** [microg/android_packages_apps_GmsCore#414]
Hi, microG works perfectly on my previous phone with LineageOS and haystack patch. However I recently changed to a new xiaomi mi6 without GMS preinstalled, and I do not intend to root or use a 3rd-party ROM.

Correct me if I'm wrong, every app that utilize GCM come with a little piece of google sdk code inside their apk. This piece of code would be very similar across all apps, and it is exactly this code asking for the signature of microG and check it against the correct one(which must be a static value inside the code). If we can decompile the and patch the target app (for instance, search/replace the static signature with microg's sig, or something more straightforward), repackage and install it, it should work with microg. The result apk will have different signature, but it can be managed.

I have over 100 app on my phone, but the ones that really need with GCM are less than 10. If we can automate this repackage process(termux have apktool so this process could be done on android), we can enjoy GCM without root/spoofing/bloatware, on any phones.

I would like to try this idea first with the search/replace trick, but I don't know google's original signature, so I can not search/replace. Could anyone give me the google signature that need to be replaced?

**Comment by WillsonChiang** [comment]
Hi are you using China ROM or Global ROM?

**Comment by ale5000-git** [comment]
You can also apply the signature spoofing patch for the ROM from the recovery (it may work without root).

**Comment by trypt0ph4n** [comment]
> Could anyone give me the google signature that need to be replaced?

I think you will find the original signature here https://github.com/thermatk/FakeGApps/blob/master/app/src/main/java/com/thermatk/android/xf/fakegapps/FakeSignatures.java

**Comment by ccaapton** [comment]
@tryptophan0 Thanks for the tip, I will try it.

@WillsonChiang I know there are ways to get GMS even on China ROM, but I do not intend to do that, it's too battery unfriendly.

@ale5000-git Thanks for your suggestion, but I don't want even unlock my bootloader. Instead I want to take a cleanest experimental approach in terms of rom-modification, and see how it goes. Wish me good luck!

**Comment by ale5000-git** [comment]
@ccaapton: You can also unlock, patch and relock (I have never tried but I suppose it works).

I don't know how to search/replace the signature but you can see the faked signature also from the sources of microG [here](https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-core/src/main/res/values/signature.xml).

**Comment by WillsonChiang** [comment]
@ccaapton Yes I have installed GMS on my China ROM and everything works perfectly except for GCM services, besides it doesn't seem to kill battery a lot.

**Comment by ale5000-git** [comment]
@WillsonChiang: My [microG unofficial installer](https://forum.xda-developers.com/showthread.php?t=3432360) support installation also on China ROMs and should fix all problems automatically.

**Comment by ale5000-git** [comment]
@ccaapton: Good start but in this way if a virus create a fake GCM then the app will happily use it, I think it would be better to check microG signature.

**Comment by chuppito** [comment]
@ccaapton  Hello is there a way to do the same for waze? Thank you
